src/nvim/screen.c

```diff
@@ -5389,66 +5247,41 @@ void screen_puts_len(char_u *text, int textlen, int row, int col, int attr)
        * (mb_off2cells() may return 2 on the right halve). */
       if (clear_next_cell)
         clear_next_cell = FALSE;
-      else if (l_has_mbyte
-               && (len < 0 ? ptr[mbyte_blen] == NUL
+      else if ((len < 0 ? ptr[mbyte_blen] == NUL
                    : ptr + mbyte_blen >= text + len)
                && ((mbyte_cells == 1 && (*mb_off2cells)(off, max_off) > 1)
                    || (mbyte_cells == 2
                        && (*mb_off2cells)(off, max_off) == 1
                        && (*mb_off2cells)(off + 1, max_off) > 1)))
         clear_next_cell = TRUE;
 
-      /* Make sure we never leave a second byte of a double-byte behind,
-       * it confuses mb_off2cells(). */
-      if (l_enc_dbcs
-          && ((mbyte_cells == 1 && (*mb_off2cells)(off, max_off) > 1)
-              || (mbyte_cells == 2
-                  && (*mb_off2cells)(off, max_off) == 1
-                  && (*mb_off2cells)(off + 1, max_off) > 1)))
-        ScreenLines[off + mbyte_blen] = 0;
       ScreenLines[off] = c;
       ScreenAttrs[off] = attr;
-      if (l_enc_utf8) {
-        if (c < 0x80 && u8cc[0] == 0)
-          ScreenLinesUC[off] = 0;
-        else {
-          int i;
+      if (c < 0x80 && u8cc[0] == 0)
+        ScreenLinesUC[off] = 0;
+      else {
+        int i;
 
-          ScreenLinesUC[off] = u8c;
-          for (i = 0; i < Screen_mco; ++i) {
-            ScreenLinesC[i][off] = u8cc[i];
-            if (u8cc[i] == 0)
-              break;
-          }
-        }
-        if (mbyte_cells == 2) {
-          ScreenLines[off + 1] = 0;
-          ScreenAttrs[off + 1] = attr;
+        ScreenLinesUC[off] = u8c;
+        for (i = 0; i < Screen_mco; ++i) {
+          ScreenLinesC[i][off] = u8cc[i];
+          if (u8cc[i] == 0)
+            break;
         }
-        screen_char(off, row, col);
-      } else if (mbyte_cells == 2) {
-        ScreenLines[off + 1] = ptr[1];
+      }
+      if (mbyte_cells == 2) {
+        ScreenLines[off + 1] = 0;
```

Comment:
What is left here is the inner body of the `if (l_enc_utf8)` starting at line 5327 (left). So the `if (mbyte_cells == 2)` here is the one at line 5340 (left), not the replacement of `else if (mbyte_cells == 2)` by `if (mbyte_cells == 2)` as suggested by the highlight.